### PR TITLE
partial build

### PR DESCRIPTION
**Pull request by fernandoblalves**

### CODE REVIEW

**Comment by acxz** [comment]
Thanks for the PR @fernandoblalves however there are still a couple things you need to do.
Please read: https://wiki.archlinux.org/index.php/PKGBUILD.
Hint: [pkgrel](https://wiki.archlinux.org/index.php/PKGBUILD#pkgrel) and [.SRCINFO](https://wiki.archlinux.org/index.php/.SRCINFO)

File: tensorflow-2.2.0.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --color --unified --recursive --text src/tensorflow-2.2.0/tensorflow/stream_executor/gpu/gpu_types.h src/tensorflow-2.2.0.new/tensorflow/stream_executor/gpu/gpu_types.h
+--- src/tensorflow-2.2.0/tensorflow/stream_executor/gpu/gpu_types.h	2020-05-05 22:58:49.000000000 +0100
++++ src/tensorflow-2.2.0.new/tensorflow/stream_executor/gpu/gpu_types.h	2020-07-14 21:07:05.406803883 +0100
+@@ -24,7 +24,7 @@
+ 
+ #include "rocm/include/hip/hip_complex.h"
+ #include "rocm/include/hip/hip_runtime.h"
+-#include "rocm/include/hiprand/hiprand.h"
++#include "rocm/hiprand/include/hiprand.h"
```
Comment (acxz) [inline tensorflow-2.2.0.patch:9]: Instead of patching the tensorflow source code it is better for us to add a symlink on the rocm-arch side. See this commit as reference: https://github.com/rocm-arch/rocm-arch/commit/b120d233c87dfd99362fa450e621367e5dbb4139
Can you make a PR to rocrand that resolves this issue so we do not need to patch tensorflow code?

File: PKGBUILD [hunk 2]
```diff
@@ -77,19 +77,34 @@ prepare() {
...
+  export HOST_C_COMPILER=/opt/rocm/hip/bin/hipcc
+  # export HOST_CXX_COMPILER=/usr/bin/g++-9
+  export HOST_CXX_COMPILER=/opt/rocm/hip/bin/hipcc
   export TF_CUDA_CLANG=0  # Clang currently disabled because it's not compatible at the moment.
   export CLANG_CUDA_COMPILER_PATH=/usr/bin/clang
   export TF_CUDA_PATHS=/opt/cuda,/usr/lib,/usr
-  export TF_CUDA_VERSION=$(/opt/cuda/bin/nvcc --version | sed -n 's/^.*release \(.*\),.*/\1/p')
-  export TF_CUDNN_VERSION=$(sed -n 's/^#define CUDNN_MAJOR\s*\(.*\).*/\1/p' /usr/include/cudnn.h)
+  export TF_CUDA_VERSION=$([ -e /opt/cuda ] && /opt/cuda/bin/nvcc --version | sed -n 's/^.*release \(.*\),.*/\1/p' || echo "")
+  export TF_CUDNN_VERSION=$([ -e /usr/include/cudnn.h ] && sed -n 's/^#define CUDNN_MAJOR\s*\(.*\).*/\1/p' /usr/include/cudnn.h || echo "")
```
Comment (acxz) [inline PKGBUILD:90]: Why is this changed? Why do we even care about it?

File: PKGBUILD [hunk 3]
```diff
@@ -77,19 +77,34 @@ prepare() {
...
+  export HCC_HOME=$ROCM_PATH/hcc
+
+  export HIP_PATH=$ROCM_PATH/hip
+  export OPENCL_ROOT=$ROCM_PATH/opencl
+  export PATH="$HCC_HOME/bin:$HIP_PATH/bin:${PATH}"
+  export PATH="$ROCM_PATH/bin:${PATH}"
+  export PATH="$OPENCL_ROOT/bin:${PATH}"
+  export HCC_HOME=/opt/rocm/hcc
+  export HIP_PATH=/opt/rocm/hip
+  export PATH=$HCC_HOME/bin:$HIP_PATH/bin:$PATH
```
Comment (acxz) [inline PKGBUILD:103]: Are the HCC env variables needed? ROCm's stack does not include hcc anymore. Those paths are not installed by ROCm packages.

File: PKGBUILD [hunk 4]
```diff
@@ -113,7 +128,7 @@ build() {
   export TF_NEED_ROCM=1
   ./configure
   bazel \
-    build --config=mkl --config=avx2_linux -c opt \
+    build -s --config=mkl --config=avx2_linux -c opt \
```
Comment (acxz) [inline PKGBUILD:131]: Im not that familiar with bazel, what does `-s` do?

**Review by acxz — CHANGES_REQUESTED** [review]

**Comment by acxz** [comment]
I tried to build it however, I got the following error: `find: '/opt/rocm/rocfft/include': No such file or directory` Should `rocm-libs` be a makedepend? Also needed `miopen-hip` installed as well, to get farther along the build.

File: tensorflow-2.2.0.patch (hunk 1 above)
Comment (fernandoblalves) [reply tensorflow-2.2.0.patch:9]: Ok, I'll try that. Should we keep that change until it is solved on their side?

File: PKGBUILD (hunk 2 above)
Comment (fernandoblalves) [reply PKGBUILD:90]: It's just to avoid an error, but I guess it is not crutial

File: PKGBUILD (hunk 3 above)
Comment (fernandoblalves) [reply PKGBUILD:103]: Perhaps I erred on the side of caution, I'll remove the unnecessary ones

File: PKGBUILD (hunk 4 above)
Comment (fernandoblalves) [reply PKGBUILD:131]: It's a verbose flag, forgot to remove it

File: tensorflow-2.2.0.patch (hunk 1 above)
Comment (acxz) [reply tensorflow-2.2.0.patch:9]: I was hoping that you would change it. Since I maintain both this repo and the rocm-arch repo I consider both projects on our side. It should be simple to implement based on the reference commit I showed you. I do not want this patch in when we merge your PR, since the problem is not here but rather in the rocrand Arch package, so we fix it there.

**Comment by fernandoblalves** [comment]
> I tried to build it however, I got the following error: `find: '/opt/rocm/rocfft/include': No such file or directory` Should `rocm-libs` be a makedepend?

I'm not sure. I have `rocm-libs` installed. The documentation on ROCm for ubuntu, RHEL, etc. states that `rocm-libs` is necessary for tensorflow.

File: PKGBUILD (hunk 2 above)
Comment (acxz) [reply PKGBUILD:90]: If it's to avoid an error that seems crucial to me, Does the build still go on? If so then lets leave it as is.

**Comment by acxz** [comment]
> The documentation ... states that rocm-libs is necessary for tensorflow.

Well there is our answer, let's add it to the makedepends.

EDIT: I released a new version with the changes to makedepends. If you could rebase your changes off of master that would be great.

File: tensorflow-2.2.0.patch (hunk 1 above)
Comment (fernandoblalves) [reply tensorflow-2.2.0.patch:9]: I believe I made a PR to fix this properly, apologies if not

Comment (acxz) [reply tensorflow-2.2.0.patch:9]: With https://github.com/rocm-arch/rocm-arch/pull/300 this patch should not be needed now. If you could remove it that would be great.

File: PKGBUILD (hunk 2 above)
Comment (fernandoblalves) [reply PKGBUILD:90]: Just to clarify, the error is that it tries to read cuda attributes and fails if you do not have cuda (which you most likely won't if you want to use rocm, if I'm not mistaken cuda is useful only if you have an nvidia gpu). The build carries on though.

Comment (acxz) [reply PKGBUILD:90]: I see kk, then it makes sense to keep this change. Can you add a comment about these two lines with a short description however?

Comment (fernandoblalves) [reply PKGBUILD:90]: Sure

**Comment by fernandoblalves** [comment]
There is one critical issue here, the original PKGBUILD uses `gcc-9` to build TF. However, I can't find it in the repos since it [seems to have been droped](https://www.phoronix.com/scan.php?page=news_item&px=Arch-GCC9-BCache-Corruption-Hit). And this is the point where I'm stuck. As you can see I replaced `gcc-9` for `hipcc` but that won't build TF's dependencies. Any suggestions? The only one I see is using `gcc-10` instead of `hipcc` for the dependencies.

Also, weirdly I'm getting these errors:
`clang++: error: unknown argument: '-fno-canonical-system-headers'`
Isn't this related to **not** using `hipcc` instead of `gcc`?

**Comment by acxz** [comment]
That is a predicament, for now the only solution is to add the `[testing]` repos to your system using gcc9 from there.

As for the clang++ errors I'm not sure.

**Comment by fernandoblalves** [comment]
I've installed `gcc-9` but I'll have to figure out how to configure bazel to use differente compilers for different parts of project. I'll pick that up in a few days.

**Comment by acxz** [comment]
@fernandoblalves closing this PR as it has gotten stale, feel free to open a new one once you make some progress.
